modular_skyrat/modules/field_of_view/code/field_of_view.dm

```diff
@@ -0,0 +1,170 @@
+
+	blocker_mask = new
+	blocker_mask.icon_state = "[fov_type]"
+	visual_shadow = new
+	visual_shadow.icon_state = "[fov_type]_v"
+	blocker_mask.dir = mob_parent.dir
+	visual_shadow.dir = mob_parent.dir
+	toggled_combat_mode(mob_parent, mob_parent.combat_mode)
+	update_fov_size()
+	add_mask()
+
+/datum/component/field_of_vision/Destroy()
+	QDEL_NULL(blocker_mask)
+	QDEL_NULL(visual_shadow)
+	return ..()
+
+/datum/component/field_of_vision/proc/update_fov_size()
+	var/mob/parent_mob = parent
+	var/client/parent_client = parent_mob.client
+	var/list/view_size = getviewsize(parent_client.view)
+	if(view_size[1] == current_fov_x && view_size[2] == current_fov_y)
+		return
+	var/x_shift = view_size[1] - current_fov_x
+	var/y_shift = view_size[2] - current_fov_y
+	var/x_scale = view_size[1] / current_fov_x
+	var/y_scale = view_size[2] / current_fov_y
+	current_fov_x = view_size[1]
+	current_fov_y = view_size[2]
+	visual_shadow.transform = blocker_mask.transform = blocker_mask.transform.Scale(x_scale, y_scale)
+	visual_shadow.transform = blocker_mask.transform = blocker_mask.transform.Translate(x_shift * 16, y_shift * 16)
+
+/datum/component/field_of_vision/proc/add_mask()
+	SIGNAL_HANDLER
+	var/mob/parent_mob = parent
+	var/client/parent_client = parent_mob.client
+	parent_client.screen += blocker_mask
+	parent_client.screen += visual_shadow
+	applied_mask = TRUE
+
+/datum/component/field_of_vision/proc/remove_mask()
+	SIGNAL_HANDLER
+	var/mob/parent_mob = parent
+	var/client/parent_client = parent_mob.client
+	parent_client.screen -= blocker_mask
+	parent_client.screen -= visual_shadow
+	applied_mask = FALSE
+
+/datum/component/field_of_vision/proc/on_dir_change(mob/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	blocker_mask.dir = new_dir
+	visual_shadow.dir = new_dir
+
+/datum/component/field_of_vision/proc/toggled_combat_mode(mob/source, new_state)
+	SIGNAL_HANDLER
+	var/mob/parent_mob = parent
+	var/target_alpha = new_state ? 255 : parent_mob.client.prefs.read_preference(/datum/preference/numeric/out_of_combat_fov_alpha)
+	visual_shadow.alpha = target_alpha
+
+/datum/component/field_of_vision/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ATOM_DIR_CHANGE, .proc/on_dir_change)
+	RegisterSignal(parent, COMSIG_LIVING_DEATH, .proc/remove_mask)
+	RegisterSignal(parent, COMSIG_LIVING_REVIVE, .proc/add_mask)
+	RegisterSignal(parent, COMSIG_LIVING_COMBAT_MODE_TOGGLE, .proc/toggled_combat_mode)
+
+/datum/component/field_of_vision/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(COMSIG_ATOM_DIR_CHANGE, COMSIG_LIVING_DEATH, COMSIG_LIVING_REVIVE, COMSIG_LIVING_COMBAT_MODE_TOGGLE))
+
+/mob/living
+	var/has_field_of_view = TRUE
+
+/mob/living/Login()
+	. = ..()
+	if(has_field_of_view && CONFIG_GET(flag/fov_enabled))
+		AddComponent(/datum/component/field_of_vision)
+
+/mob/living/Logout()
+	var/fov_component = GetComponent(/datum/component/field_of_vision)
+	if(fov_component)
+		qdel(fov_component)
+	. = ..()
```

Comment:
why not hook this onto `COMSIG_MOB_LOGOUT` on the component itself

Reply:
it made sense for me since the application of it was in Login, I suppose I could move it